File: network_cs/Networking/jsonprotocol/FestivalClientWorker.cs
using log4net;
using Services;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using Dtos;
using System.Runtime.CompilerServices;

namespace Networking.jsonprotocol
{
    internal class FestivalClientWorker : IFestivalObserver
    {
        private IFestivalService server;
        private TcpClient connection;

        private NetworkStream stream;
        private volatile bool connected;
        private static readonly ILog log = LogManager.GetLogger(typeof(FestivalClientWorker));
        public FestivalClientWorker(IFestivalService server, TcpClient connection)
        {
            this.server = server;
            this.connection = connection;
            try
            {

                stream = connection.GetStream();
                connected = true;
            }
            catch (Exception e)
            {
                log.Error(e.StackTrace);
            }
        }

        private Response handleRequest(Request request)
        {
            Response response = null;
            try
            {
                switch (request.type)
                {
                    case RequestType.LOGIN:
                        return handleLogin(request);
                    case RequestType.GET_ALL_CONCERTS:
                        return handleGetAllConcerts(request);
                    case RequestType.SELL_TICKETS:
                        return handleSellTickets(request);
                    case RequestType.LOGOUT:
                        return handleLogout(request);
                }
            }
            catch (Exception e)
            {
                log.Error(e.StackTrace);
            }
            return response;
        }

        private Response handleGetAllConcerts(Request request)
        {
            Response response = null;
            try
            {
                var concerts = server.GetConcerts();
                response = new Response.Builder()
                    .Type(ResponseType.GET_CONCERTE)
                    .Concerts(concerts)
                    .Build();
            }
            catch (Exception e)
            {
                response = new Response.Builder()
                    .Type(ResponseType.ERROR)
                    .Message(e.Message)
                    .Build();
            }
            return response;
        }

        private Response handleLogin(Request request)
        {
            Response response = null;
            try
            {
                server.Login(request.angajat, this);
                response = new Response.Builder()
                    .Type(ResponseType.OK)
                    .Message("Login successful")
                    .Build();
            }
            catch (Exception e)
            {
                response = new Response.Builder()
                    .Type(ResponseType.ERROR)
                    .Message(e.Message)
                    .Build();
            }
            return response;
        }

        private Response handleSellTickets(Request request)
        {
            Response response = null;
            try
            {
                int nrTickets = request.nrTickets ?? 0;
                server.SellTickets(request.concert, nrTickets);
                response = new Response.Builder()
                    .Type(ResponseType.OK)
                    .Message("Tickets sold successfully")
                    .Build();
            }
            catch (Exception e)
            {
                response = new Response.Builder()
                    .Type(ResponseType.ERROR)
                    .Message(e.Message)
                    .Build();
            }
            return response;
        }

        private Response handleLogout(Request request)
        {
            Response response = null;
            try
            {
                server.Logout(request.angajat);
                connected = false;
                response = new Response.Builder()
                    .Type(ResponseType.OK)
                    .Message("Logout successful")
                    .Build();
            }
            catch (Exception e)
            {
                log.Error(e.StackTrace);
                response = new Response.Builder()
                    .Type(ResponseType.ERROR)
                    .Message(e.Message)
                    .Build();
            }
            return response;
        }
        public virtual void run()
        {
            using StreamReader reader = new StreamReader(stream, Encoding.UTF8);
            while (connected)
            {
                try
                {
                    string requestJson = reader.ReadLine(); // Read JSON line-by-line
                    if (string.IsNullOrEmpty(requestJson)) continue;
                    log.DebugFormat("Received json request {0}", requestJson);
                    Request request = JsonSerializer.Deserialize<Request>(requestJson);
                    log.DebugFormat("Deserializaed Request {0} ", request);
                    Response response = handleRequest(request);
                    if (response != null)
                    {
                        sendResponse(response);
                    }
                }
                catch (Exception e)
                {
                    log.ErrorFormat("run eroare {0}", e.Message);
                    if (e.InnerException != null)
                        log.ErrorFormat("run inner error {0}", e.InnerException.Message);
                    log.Error(e.StackTrace);
                }

                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    log.Error(e.StackTrace);
                }
            }
            try
            {
                stream.Close();
                connection.Close();
            }
            catch (Exception e)
            {
                log.Error("Error " + e);
            }
        }

        private void sendResponse(Response response)
        {
            //de modificat pentru json
            String jsonString = JsonSerializer.Serialize(response);
            log.DebugFormat("sending response {0}", jsonString);
            lock (stream)
            {
                byte[] data = Encoding.UTF8.GetBytes(jsonString + "\n"); // Append newline 
                stream.Write(data, 0, data.Length);
                stream.Flush();
            }

        }

        public void UpdateConcert(ConcertDto c)
        {
            Response response = new Response.Builder()
                .Type(ResponseType.UPDATE_CONCERT)
                .Concert(c)
                .Build();
            try
            {
                sendResponse(response);
            }
            catch (Exception e)
            {
                log.ErrorFormat("Error sending update {0}", e.Message);
                if (e.InnerException != null)
                    log.ErrorFormat("Error sending update inner {0}", e.InnerException.Message);
            }
        }
    }
}

File: network_java/FestivalNetworking/src/main/java/networking/rpcprotocol/FestivalClientWorker.java
package networking.rpcprotocol;

import dtos.AngajatDto;
import dtos.BiletDto;
import dtos.ConcertDto;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import services.FestivalException;
import services.IFestivalObserver;
import services.IFestivalService;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.Socket;

public class FestivalClientWorker implements Runnable , IFestivalObserver {
    private IFestivalService server;
    private Socket connection;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;

    private static Logger logger = LogManager.getLogger(FestivalClientWorker.class);

    private static Response okResponse=new Response.Builder().type(ResponseType.OK).build();

    public FestivalClientWorker(IFestivalService server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try{
            output=new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input=new ObjectInputStream(connection.getInputStream());
            connected=true;
        } catch (IOException e) {
            logger.error(e);
            logger.error(e.getStackTrace());
        }
    }

    public void run() {
        while(connected){
            try {
                Object request=input.readObject();
                Response response=handleRequest((Request)request);
                if (response!=null){
                    sendResponse(response);
                }
            } catch (IOException|ClassNotFoundException e) {
                logger.error(e);
                logger.error(e.getStackTrace());
            }
            try {
                Thread.sleep(100);
            } catch (InterruptedException e) {
                logger.error(e);
                logger.error(e.getStackTrace());
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            logger.error("Error "+e);
        }
    }
    private Response handleRequest(Request request){
        Response response=null;
        String handlerName="handle"+(request).getType();
        logger.debug("HandlerName "+handlerName);
        try {
            Method method=this.getClass().getDeclaredMethod(handlerName, Request.class);
            response=(Response)method.invoke(this,request);
            logger.debug("Method "+handlerName+ " invoked");
        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
            logger.error(e);
            logger.error(e.getStackTrace());
        }

        return response;
    }

    private Response handleLOGOUT(Request request){
        logger.debug("Handle LOGOUT");
        AngajatDto angajat = (AngajatDto)request.getData();
        try{
            server.logout(angajat);
            connected=false;
            return okResponse;
        } catch (FestivalException e) {
            logger.error(e);
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }

    }

    private Response handleLOGIN(Request request){
        logger.debug("Login request ..."+request.getType());
        AngajatDto udto=(AngajatDto) request.getData();
        try {
            server.login(udto.getNume(), udto.getPassword(), this);
            return okResponse;
        } catch (FestivalException e) {
            connected=false;
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleGET_CONCERTE(Request request){
        logger.debug("Get concerte request ..."+request.getType());
        try{
            return new Response.Builder().type(ResponseType.GET_CONCERTE).data(server.getConcerts()).build();
        } catch (FestivalException e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleVANZARE(Request request){
        logger.debug("Vanzare request ..."+request.getType());
        try{
            BiletDto bilet = (BiletDto) request.getData();
            server.sellTickets(bilet.getConcert(), bilet.getNrLocuri());
            return okResponse;
        } catch (FestivalException e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private void sendResponse(Response response) throws IOException{
        logger.debug("sending response {}",response);
        synchronized (output) {
            output.writeObject(response);
            output.flush();
        }
    }


    @Override
    public void updateBilete(ConcertDto concert) throws FestivalException {
        logger.debug("updateBilete concert ..."+concert);
        Response resp =new Response.Builder().type(ResponseType.UPDATE_CONCERT).data(concert).build();
        try{
            sendResponse(resp);
        } catch (IOException e){
            logger.error(e);
            logger.error(e.getStackTrace());
        }
    }
}
